src/dp/distributions.rs

```diff
@@ -0,0 +1,340 @@
+    buffer[0] & 1 == 1
+}
+
+/// sample from the Bernoulli(gamma) distribution, where $gamma /leq 1$.
+///
+/// `sample_bernoulli_frac(gamma, rng)` returns numbers distributed as $Bernoulli(gamma)$.
+/// using the given random number generator for base randomness.
+fn sample_bernoulli_frac<R: Rng + ?Sized>(gamma: &Ratio<BigUint>, rng: &mut R) -> bool {
+    let d = gamma.denom();
+    assert!(!d.is_zero());
+    assert!(gamma <= &Ratio::<BigUint>::one());
+
+    // sample uniform biguint in [0,d)
+    let s = rng.gen_range(BigUint::zero()..d.clone());
+
+    s < *gamma.numer()
+}
+
+/// sample from the Bernoulli(exp(-gamma)) distribution, where $gamma \leq 1$.
+///
+/// `sample_bernoulli_exp1(gamma, rng)` returns numbers distributed as $Bernoulli(exp(-gamma))$.
+/// using the given random number generator for base randomness.
+fn sample_bernoulli_exp1<R: Rng + ?Sized>(gamma: &Ratio<BigUint>, rng: &mut R) -> bool {
+    assert!(!gamma.denom().is_zero());
+    assert!(gamma <= &Ratio::<BigUint>::one());
+    let mut k = BigUint::one();
+    loop {
+        if sample_bernoulli_frac(&(gamma / k.clone()), rng) {
+            k += 1u8;
+        } else {
+            return !(k % BigUint::from(2u8)).is_zero();
+        }
+    }
+}
+
+/// sample from the Bernoulli(exp(-gamma)) distribution.
+///
+/// `sample_bernoulli_exp(gamma, rng)` returns numbers distributed as $Bernoulli(exp(-gamma))$,
+/// using the given random number generator for base randomness.
+fn sample_bernoulli_exp<R: Rng + ?Sized>(gamma: &Ratio<BigUint>, rng: &mut R) -> bool {
+    assert!(!gamma.denom().is_zero());
+    // sample floor(n/d) independent Bernoulli(exp(-1))
+    // If all are 1, return Bernoulli(exp(-(gamma-floor(gamma))))
+    let mut gamma: Ratio<BigUint> = gamma.clone();
+    while Ratio::<BigUint>::one() < gamma {
+        if !sample_bernoulli_exp1(&Ratio::<BigUint>::one(), rng) {
+            return false;
+        }
+        gamma -= Ratio::<BigUint>::one();
+    }
+    sample_bernoulli_exp1(&gamma, rng)
+}
+
+/// sample from the geometric distribution with parameter 1 - exp(-gamma) (slow).
+///
+/// `sample_geometric_exp_slow(gamma, rng)` returns numbers distributed according to
+/// $Geometric(1 - exp(-gamma))$, using the given random number generator for base randomness.
+fn sample_geometric_exp_slow<R: Rng + ?Sized>(gamma: &Ratio<BigUint>, rng: &mut R) -> BigUint {
+    assert!(!gamma.denom().is_zero());
+    let mut k = BigUint::zero();
+    loop {
+        if sample_bernoulli_exp(gamma, rng) {
+            k += 1u8;
+        } else {
+            return k;
+        }
+    }
+}
+
+/// sample from the geometric distribution  with parameter 1 - exp(-gamma) (fast).
+///
+/// `sample_geometric_exp_fast(gamma, rng)` returns numbers distributed according to
+/// $Geometric(1 - exp(-gamma))$, using the given random number generator for base randomness.
+fn sample_geometric_exp_fast<R: Rng + ?Sized>(gamma: &Ratio<BigUint>, rng: &mut R) -> BigUint {
```

Comment:
Update: it turns out that the reference implementation (https://github.com/IBM/discrete-gaussian-differential-privacy/blob/master/discretegauss.py), written by the CKS20 authors, differs from the pseudocode of their own paper. I'll try to understand why.

Reply:
It seems fine to match the reference implementation, but we should document the changes from the paper because it could end up being important for a security audit. WDYT @ooovi, @MxmUrw?

Reply:
i think the reason for diverging here is performance; it's equivalent code avoiding recursive calls. i reimplemented following the pseudocode and will run some benchmarks to see how good of a reason that is -- as this will run on big numbers, we might need the optimization. i can't really tell how important sticking to the pseudocode is for auditors, but i can def document all the changes

Reply:
It makes perfect sense to make a recursive algorithm iterative.